File: python/flink_agents/api/trace/execution_reporter.py
################################################################################
#  Licensed to the Apache Software Foundation (ASF) under one
#  or more contributor license agreements.  See the NOTICE file
#  distributed with this work for additional information
#  regarding copyright ownership.  The ASF licenses this file
#  to you under the Apache License, Version 2.0 (the
#  "License"); you may not use this file except in compliance
#  with the License.  You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
# limitations under the License.
#################################################################################
import logging
from abc import ABC, abstractmethod
from collections.abc import Callable, Mapping
from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from flink_agents.api.runner_context import RunnerContext

logger = logging.getLogger(__name__)

_EMPTY_METADATA: Mapping[str, Any] = {}


class ExecutionEntityTypes:
    """Shared entity type names for execution reports."""

    ACTION = "action"
    LLM = "llm"
    PARSER = "parser"
    TOOL = "tool"


class ExecutionProblemCategories:
    """Shared low-cardinality problem categories for failed execution reports."""

    ACTION_EXECUTION_FAILED = "action_execution_failed"
    MODEL_CALL_FAILED = "model_call_failed"
    MODEL_OUTPUT_PARSE_ERROR = "model_output_parse_error"
    TOOL_CALL_FAILED = "tool_call_failed"


class ExecutionReporter(ABC):
    """Optional capability for reporting executions nested inside an action."""

    @abstractmethod
    def report_execution_started(
        self,
        entity_type: str,
        entity_name: str,
        entity_metadata: Mapping[str, Any] | None = None,
    ) -> None:
        """Report that a logical execution started."""

    @abstractmethod
    def report_execution_succeeded(
        self,
        entity_type: str,
        entity_name: str,
        entity_metadata: Mapping[str, Any] | None = None,
    ) -> None:
        """Report that a logical execution completed successfully."""

    @abstractmethod
    def report_execution_failed(
        self,
        entity_type: str,
        entity_name: str,
        entity_metadata: Mapping[str, Any] | None,
        error: BaseException,
        problem_category: str | None = None,
    ) -> None:
        """Report that a logical execution failed."""


class ExecutionReporters:
    """Best-effort helpers for contexts that implement ExecutionReporter."""

    @staticmethod
    def started(
        ctx: "RunnerContext",
        entity_type: str,
        entity_name: str,
        entity_metadata: Mapping[str, Any] | None = None,
    ) -> None:
        """Report the start of a nested execution if the context supports it."""
        ExecutionReporters._report(
            ctx,
            lambda reporter: reporter.report_execution_started(
                entity_type, entity_name, entity_metadata or _EMPTY_METADATA
            ),
        )

    @staticmethod
    def succeeded(
        ctx: "RunnerContext",
        entity_type: str,
        entity_name: str,
        entity_metadata: Mapping[str, Any] | None = None,
    ) -> None:
        """Report successful completion if the context supports it."""
        ExecutionReporters._report(
            ctx,
            lambda reporter: reporter.report_execution_succeeded(
                entity_type, entity_name, entity_metadata or _EMPTY_METADATA
            ),
        )

    @staticmethod
    def failed(
        ctx: "RunnerContext",
        entity_type: str,
        entity_name: str,
        entity_metadata: Mapping[str, Any] | None,
        error: BaseException,
        problem_category: str | None = None,
    ) -> None:
        """Report failed completion if the context supports it."""
        ExecutionReporters._report(
            ctx,
            lambda reporter: reporter.report_execution_failed(
                entity_type,
                entity_name,
                entity_metadata or _EMPTY_METADATA,
                error,
                problem_category,
            ),
        )

    @staticmethod
    def _report(
        ctx: "RunnerContext", report: Callable[[ExecutionReporter], None]
    ) -> None:
        if not isinstance(ctx, ExecutionReporter):
            return
        try:
            report(ctx)
        except Exception:
            logger.debug("Execution reporting failed and was ignored.", exc_info=True)

File: api/src/main/java/org/apache/flink/agents/api/trace/ExecutionReporter.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.flink.agents.api.trace;

import javax.annotation.Nullable;

import java.util.Map;

/**
 * Optional capability for reporting logical executions nested inside the current action.
 *
 * <p>Implementations decide how reports are consumed or ignored. Callers should provide stable
 * entity type/name pairs and keep metadata small, structured, serializable, and stable for equality
 * matching between lifecycle reports of the same logical execution.
 */
public interface ExecutionReporter {

    /** Shared entity type names for execution reports. */
    final class EntityTypes {
        public static final String ACTION = "action";
        public static final String LLM = "llm";
        public static final String PARSER = "parser";
        public static final String TOOL = "tool";

        private EntityTypes() {}
    }

    /** Shared low-cardinality problem categories for failed execution reports. */
    final class ProblemCategories {
        public static final String ACTION_EXECUTION_FAILED = "action_execution_failed";
        public static final String MODEL_CALL_FAILED = "model_call_failed";
        public static final String MODEL_OUTPUT_PARSE_ERROR = "model_output_parse_error";
        public static final String TOOL_CALL_FAILED = "tool_call_failed";

        private ProblemCategories() {}
    }

    /**
     * Reports that a logical execution has been created but has not necessarily started.
     *
     * <p>This is an optional lifecycle phase for executions whose admission and invocation are
     * observably separate. Implementations that do not consume it may keep the default no-op. A
     * later start or terminal report is not guaranteed, so consumers must not infer whether the
     * underlying invocation ran from the absence of either report.
     *
     * @param entityType stable category of the reported execution, such as LLM, parser, or tool
     * @param entityName stable name of the reported execution, such as model or tool name
     * @param entityMetadata small structured metadata used to match subsequent lifecycle reports
     */
    default void reportExecutionCreated(
            String entityType, String entityName, Map<String, Object> entityMetadata)
            throws Exception {}

    /**
     * Reports that a logical execution started within the current action.
     *
     * @param entityType stable category of the reported execution, such as LLM, parser, or tool
     * @param entityName stable name of the reported execution, such as model or tool name
     * @param entityMetadata small structured metadata used to distinguish repeated executions with
     *     the same type/name
     */
    void reportExecutionStarted(
            String entityType, String entityName, Map<String, Object> entityMetadata)
            throws Exception;

    /**
     * Reports that a logical execution started at the given occurrence timestamp.
     *
     * <p>The default implementation delegates to {@link #reportExecutionStarted(String, String,
     * Map)}, so reporters that do not retain occurrence timestamps may use their observation time.
     */
    default void reportExecutionStartedAt(
            String entityType,
            String entityName,
            Map<String, Object> entityMetadata,
            String timestamp)
            throws Exception {
        reportExecutionStarted(entityType, entityName, entityMetadata);
    }

    /**
     * Reports that a logical execution completed successfully.
     *
     * <p>The entity type/name/metadata should match the corresponding start report when one was
     * reported.
     */
    void reportExecutionSucceeded(
            String entityType, String entityName, Map<String, Object> entityMetadata)
            throws Exception;

    /**
     * Reports that a logical execution completed successfully at the given occurrence timestamp.
     *
     * <p>The default implementation delegates to {@link #reportExecutionSucceeded(String, String,
     * Map)}, so reporters that do not retain occurrence timestamps may use their observation time.
     */
    default void reportExecutionSucceededAt(
            String entityType,
            String entityName,
            Map<String, Object> entityMetadata,
            String timestamp)
            throws Exception {
        reportExecutionSucceeded(entityType, entityName, entityMetadata);
    }

    /**
     * Reports that a logical execution failed.
     *
     * <p>The entity type/name/metadata should match the corresponding start report when one was
     * reported. The problem category should be a stable, low-cardinality classification.
     */
    void reportExecutionFailed(
            String entityType,
            String entityName,
            Map<String, Object> entityMetadata,
            Throwable error,
            @Nullable String problemCategory)
            throws Exception;

    /**
     * Reports that a logical execution failed at the given occurrence timestamp.
     *
     * <p>The default implementation delegates to {@link #reportExecutionFailed(String, String, Map,
     * Throwable, String)}, so reporters that do not retain occurrence timestamps may use their
     * observation time.
     */
    default void reportExecutionFailedAt(
            String entityType,
            String entityName,
            Map<String, Object> entityMetadata,
            Throwable error,
            @Nullable String problemCategory,
            String timestamp)
            throws Exception {
        reportExecutionFailed(entityType, entityName, entityMetadata, error, problemCategory);
    }
}
